shipyard_blueprints/athena/shipyard_athena/athena.py

```diff
@@ -1,35 +1,224 @@
+        logger.debug("Started query execution")
+        job_id = job["QueryExecutionId"]
+        logger.debug(f"Fetched job ID {job_id}")
+        status = self._get_query_status(job_id)
+        logger.debug(f"Query status is {status}")
+        while not status:
+            time.sleep(5)
+            logger.debug("Wating another 5 seconds to check query status")
+            status = self._get_query_status(job_id)
+            logger.debug(f"Query status is {status}")
+
+        logger.debug(f"Status JSON reads: {status}")
+        return status["QueryExecution"]["Status"]["State"]
+
+    def fetch(
+        self,
+        query: str,
+        dest_path: str,
+        database: Optional[str] = None,
+        log_folder: Optional[str] = None,
+    ):
+        """
+        Fetches the result of a query from Athena and saves it to a destination path.
+
+        Args:
+            query (str): The SQL query to execute.
+            dest_path (str): The path where the query result will be saved.
+            database (Optional[str]): The name of the database to execute the query in. Defaults to None.
+            log_folder (Optional[str]): The folder where query logs are stored. Defaults to None.
+
+        Returns:
+            str: The response from the download operation.
 
-    def execute_query(self, query: str):
-        pass
+        Raises:
+            ExitCodeException: If an exit code exception occurs.
+            errs.FetchError: If any other exception occurs during the fetch operation.
+        """
+        try:
+            job = self._execute_query(
+                query=query, database=database, log_folder=log_folder
+            )
+            job_id = job["QueryExecutionId"]
+            status = self._get_query_status(job_id)
+            logger.debug(f"Query status is {status}")
+            while not status:
+                time.sleep(5)
+                logger.debug("Waiting another 5 seconds to check query status")
+                status = self._get_query_status(job_id)
+                logger.debug(f"Query status is {status}")
+
+            response = self.s3.Bucket(self.bucket).download_file(
+                f'{log_folder}{"/" if log_folder else ""}{job_id}.csv', dest_path
+            )
+            logger.debug("Download complete")
+            if response:
+                logger.debug(f"Contents of response: {response}")
+        except ExitCodeException:
+            raise
+        except Exception as e:
+            raise errs.FetchError(e)
+        else:
+            return response
+
+    def _get_query_status(self, job_id: str):
+        """Fetches the query status
 
-    def fetch(self, query: str):
-        pass
+        Args:
+            job_id: The ID of the associated query to fetch
+        """
+        result = self.athena.get_query_execution(QueryExecutionId=job_id)
+        state = result["QueryExecution"]["Status"]["State"]
+        if state == "SUCCEEDED":
```

Comment:
This is being used in a while loop which always makes me nervous.
* Are we certain that `SUCCEEDED` is the only value it should exit the loop for? Are there not values like timeout or canceled
* If possible we should handle all known states and raise an error on unknown state to make ideally with its own exit code so that we can easily add the logic to handle the future states otherwise it would be difficult to pinpoint the issue if the loop never exits

Reply:
thoughts on the above?

Reply:
yes - missed this for some reason on my last push. I need to handle for queued, cancelled, and running (those are all the possible statuses) [](edited-with-Axolo)